pkg/cloudprovider/overlay/cloudprovider.go

```diff
@@ -0,0 +1,197 @@
+	return &decorator{CloudProvider: cloudProvider, kubeClient: kubeClient}
+}
+
+func (d *decorator) GetInstanceTypes(ctx context.Context, nodePool *v1.NodePool) ([]*cloudprovider.InstanceType, error) {
+	its, err := pullInstanceTypes(ctx, d.CloudProvider, nodePool)
+	if err != nil {
+		return []*cloudprovider.InstanceType{}, err
+	}
+	if options.FromContext(ctx).FeatureGates.NodeOverlay {
+		// The additional requirements will be added to the instance type during scheduling simulation
+		// Since getting instance types is done on a NodePool level, these requirnements were always assumed
+		// to be allowed with these instance types.
+		addNodePoolRequirements(nodePool, its)
+		its, err = OverlayInstanceTypes(ctx, d.kubeClient, its)
+		if err != nil {
+			return []*cloudprovider.InstanceType{}, err
+		}
+		removeNodePoolRequirements(nodePool, its)
+	}
+	return its, nil
+}
+
+func OverlayInstanceTypes(ctx context.Context, kubeClient client.Client, instanceTypes []*cloudprovider.InstanceType) ([]*cloudprovider.InstanceType, error) {
+	overlays := &v1alpha1.NodeOverlayList{}
+	err := kubeClient.List(ctx, overlays)
+	if err != nil {
+		return []*cloudprovider.InstanceType{}, err
+	}
+
+	// Only apply overlays that have passed runtime validation and know there have not been
+	// any updates to the spec of the overlay field
+	validOverlays := lo.Filter(overlays.Items, func(o v1alpha1.NodeOverlay, _ int) bool {
+		condition := o.StatusConditions().Get(v1alpha1.ConditionTypeValidationSucceeded)
+		return condition.IsTrue() && condition.ObservedGeneration == o.Generation
+	})
+
+	// We order the overlays from largest to smallest, as we have already done validation
+	// we can be sure that we will not be blocked due to invalid or conflicting overlays
+	sort.Slice(validOverlays, func(x int, y int) bool {
+		return lo.FromPtr(validOverlays[x].Spec.Weight) > lo.FromPtr(validOverlays[y].Spec.Weight)
+	})
+	work := []func(overlays []v1alpha1.NodeOverlay, its []*cloudprovider.InstanceType){
+		overlayCapacityOnInstanceTypes,
+		overlayPriceOnInstanceTypes,
+	}
+	lop.ForEach(work, func(f func(overlays []v1alpha1.NodeOverlay, its []*cloudprovider.InstanceType), i int) {
+		f(validOverlays, instanceTypes)
+	})
+
+	return instanceTypes, nil
+}
+
+func overlayCapacityOnInstanceTypes(overlays []v1alpha1.NodeOverlay, its []*cloudprovider.InstanceType) {
+	for _, overlay := range overlays {
+		overlaySelector := scheduling.NewRequirements()
+		overlaySelector.Add(scheduling.NewNodeSelectorRequirements(overlay.Spec.Requirements...).Values()...)
+
+		for _, it := range its {
+			if it.Requirements.IsCompatible(overlaySelector) {
+				it.Capacity = lo.Assign(it.Capacity, overlay.Spec.Capacity)
+				it.ApplyResourceOverlay()
+			}
+		}
+	}
+}
+
+func overlayPriceOnInstanceTypes(overlays []v1alpha1.NodeOverlay, its []*cloudprovider.InstanceType) {
+	// we will only update instance types once
+	overriddenInstanceType := map[string][]string{}
```

Comment:
Why do we need this check here, but not for capacity. In both cases shouldn't validation have ensured that there aren't any conflicts?

Reply:
Validation handles cases for clear conflicts where the weight is not defined, however for weight difference we will only apply at the overlays with the higher weight. I have updated this function to drop the overlay capacity function, so this does not matter anymore